### PR TITLE
Sized AbstractArray

### PR DESCRIPTION
**Pull request by mateuszbaran**
Allocation-free statically sized views in Julia 1.5 🎉

The code should be more-or-less complete but I need to add more tests.

One thing that is potentially controversial (and subject to change) is allowing one way of reshaping. Sometimes I represent arrays of different sizes as parts of a long 1-dimensional array, I could of course call `reshape` before wrapping in `SizedArray` but that's easy enough to do directly in `SizedArray`, and currently `SizedArray` allows for even more reshaping. Relevant PR: #666.

This PR supersedes #341 .

### CODE REVIEW

**Comment by c42f** [comment]
The CI failures here are weird - it's a git error at checkout where supposedly  d2adcd0706644b3fab095bb0b2fe9c1080bea20f doesn't exist. But it exists in my local repo.

**Comment by mateuszbaran** [comment]
I've pushed a new commit and the tests seem to be running OK.

I have one question here: do you think `view(::SizedArray, something...)` should (where possible) return a `SizedArray` wrapping a view of the wrapped `AbstractArray`?

**Comment by andyferris** [comment]
> I have one question here: do you think view(::SizedArray, something...) should (where possible) return a SizedArray wrapping a view of the wrapped AbstractArray?

To me - the ideal behavior has `axes(getindex(a, b)) === axes(view(a, b))`.

If this involves effort I would wait until we support `axes` which are "partially static" (a mixture of `OneTo` and `SOneTo`) and we can clean it all up then.

**Comment by mateuszbaran** [comment]
> To me - the ideal behavior has `axes(getindex(a, b)) === axes(view(a, b))`.

As far as I can tell, this is already the case 🙂 . I'm primarily concerned that without swapping of `SubArray` and `SizedArray` when taking a `view` lots of methods from `StaticArrays.jl` couldn't be used.

**Comment by mateuszbaran** [comment]
I think this is ready for review now. I have finished `view`s of `SizedArray` and `MArray`.

File: src/SizedArray.jl [hunk 1]
```diff
@@ -53,43 +64,156 @@ end
...
+    return Array(reshape(sa.data, size_to_tuple(S)))
+end
+@inline function Base.Array{T,N}(sa::SizedArray{S,T,N}) where {T,S,N}
+    return Array(reshape(sa.data, size_to_tuple(S)))
+end
 
-@inline convert(::Type{Array}, sa::SizedArray) = sa.data
-@inline convert(::Type{Array{T}}, sa::SizedArray{S,T}) where {T,S} = sa.data
-@inline convert(::Type{Array{T,N}}, sa::SizedArray{S,T,N}) where {T,S,N} = sa.data
+@inline function convert(::Type{Array}, sa::SizedArray{S}) where {S}
```
Comment (andyferris) [inline src/SizedArray.jl:95]: This only valid if `sa.data isa Array` - so we'll need a fallback that converts if not.

Comment (andyferris) [reply src/SizedArray.jl:95]: Oh never mind I see you did it below!

File: src/SizedArray.jl [hunk 2]
```diff
@@ -53,43 +64,156 @@ end
...
+    a::SizedArray{S},
+    indices::Union{Integer, Colon, StaticVector, Base.Slice, SOneTo}...,
+) where {S}
+    new_size = new_out_size(S, indices...)
+    return SizedArray{new_size}(view(a.data, indices...))
+end
 
-function promote_rule(::Type{<:SizedArray{S,T,N,M}}, ::Type{<:SizedArray{S,U,N,M}}) where {S,T,U,N,M}
-    SizedArray{S,promote_type(T,U),N,M}
+function Base.view(
```
Comment (andyferris) [inline src/SizedArray.jl:212]: We really need to move this to the MArray.jl file.

Comment (mateuszbaran) [reply src/SizedArray.jl:212]: Sure, I'll move it.

Comment (mateuszbaran) [reply src/SizedArray.jl]: I think this can be marked as resolved?

**Comment by mateuszbaran** [comment]
I've fixed one remaining problem with views. Failure on nightly is related to a change in printing of types.

**Comment by chriselrod** [comment]
Could you define `parent` now that the `parent` array's properties determine those of the `SizedArray`?

**Comment by mateuszbaran** [comment]
Sure, I've just added `parent` (and `vec`).

**Comment by mateuszbaran** [comment]
Are there any plans to release StaticArrays 0.13 sometime soonish? If not, I think I'll put sized abstract arrays in HybridArrays.jl for now. It would solve many problems with `SSubArray`.

**Comment by c42f** [comment]
@mateuszbaran I approved this, so I was kind of waiting for you to go ahead and merge it. Go on :)

Regarding release, I was imagining we should just release 1.0 next, following the most conservative release plan. IIRC there was one other minor breaking change we should do.

**Comment by mateuszbaran** [comment]
Great! I will merge it.

Releasing 1.0 would be quite a big thing and I'm happy with your plan about it :+1: .

**Pull request merged**